tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java

```diff
@@ -270,36 +273,37 @@ public void stop(BundleContext context) throws Exception {
 		super.stop(context);
 	}
 
-	private static <T extends ILiferayProject> T _checkProjectCache(Class<T> type, Object adaptable) {
+	private static <T extends ILiferayProject> ILiferayProject _checkProjectCache(Class<T> type, Object adaptable) {
 		Map<ProjectCacheKey<?>, ILiferayProject> projectCache = _plugin._projectCache;
 
 		ProjectCacheKey<T> projectCacheKey = new ProjectCacheKey<>(type, adaptable);
 
 		ILiferayProject liferayProject = projectCache.get(projectCacheKey);
 
-		if (liferayProject == null) {
-			for (ILiferayProject cachedLiferayProject : projectCache.values()) {
-				if (type.isInstance(cachedLiferayProject) && adaptable.equals(cachedLiferayProject.getProject())) {
-					liferayProject = type.cast(cachedLiferayProject);
-
-					break;
-				}
-			}
-		}
+		//Stale project should be removed in the first time
 
 		if ((liferayProject != null) && liferayProject.isStale()) {
-			if (liferayProject instanceof EventListener) {
-				ListenerRegistry listenerRegistry = listenerRegistry();
+			liferayProject = null;
+		}
 
-				listenerRegistry.removeEventListener((EventListener)liferayProject);
-			}
+		//Handle the situation that the adaptable object could have muti-type
 
-			projectCache.remove(projectCacheKey);
+		if (liferayProject == null) {
+			for (Entry<ProjectCacheKey<?>, ILiferayProject> entry : projectCache.entrySet()) {
+				ILiferayProject cachedLiferayProject = entry.getValue();
 
-			liferayProject = null;
+				if (type.isInstance(cachedLiferayProject) && adaptable.equals(cachedLiferayProject.getProject())) {
+					if (cachedLiferayProject.isStale()) {
+						_removeFromCache(entry);
```

Comment:
Hey @gamerson , this will get into trouble with `ConcurrentModificationException` if you try to remove key when iterating it. and i think my previous change would be better that remove the cache in first `get()`.
My original idea:
1. Try to get entry from cache if the key is 100 percent matched, if not , we need remove it 
2. Try to get entry from cache if the key may be 'matched', in this opertaion, the stale key will be skiped and leave ther e until to the next `_checkProjectCache()` come, so this stale cache can be removed by the 100 percent matched key.